Firestore/core/test/unit/model/values_test.cc

```diff
@@ -0,0 +1,323 @@
+        serializer.EncodeResourceName(database_id, key.path());
+    return result;
+  }
+
+  google_firestore_v1_Value WrapServerTimestamp(
+      const model::FieldValue& input) {
+    // TODO(mrschmidt): Replace with EncodeFieldValue encoding when available
+    return WrapObject("__type__", "server_timestamp", "__local_write_time__",
+                      input.server_timestamp_value().local_write_time());
+  }
+
+  template <typename... Args>
+  void Add(std::vector<std::vector<google_firestore_v1_Value>>& groups,
+           Args... values) {
+    std::vector<google_firestore_v1_Value> group{(values)...};
+    groups.emplace_back(group);
+  }
+
+  void VerifyEquals(std::vector<google_firestore_v1_Value>& group) {
+    for (size_t i = 0; i < group.size(); ++i) {
+      for (size_t j = 0; j < group.size(); ++j) {
+        EXPECT_TRUE(Values::Equals(group[i], group[j]));
+      }
+    }
+  }
+
+  void VerifyNotEquals(std::vector<google_firestore_v1_Value>& left,
+                       std::vector<google_firestore_v1_Value>& right) {
+    for (const auto& val1 : left) {
+      for (const auto& val2 : right) {
+        EXPECT_FALSE(Values::Equals(val1, val2));
+      }
+    }
+  }
+
+  void VerifyOrdering(std::vector<google_firestore_v1_Value>& left,
+                      std::vector<google_firestore_v1_Value>& right,
+                      ComparisonResult cmp) {
+    for (const auto& val1 : left) {
+      for (const auto& val2 : right) {
+        EXPECT_EQ(cmp, Values::Compare(val1, val2));
+      }
+    }
+  }
+
+  void VerifyCanonicalId(const google_firestore_v1_Value& value,
+                         const std::string& expected_canonical_id) {
+    const std::string& actual_canonical_id = Values::CanonicalId(value);
+    EXPECT_EQ(expected_canonical_id, actual_canonical_id);
+  }
+
+ private:
+  remote::Serializer serializer;
+};
+
+TEST_F(ValuesTest, Equality) {
+  std::vector<std::vector<google_firestore_v1_Value>> equals_group;
+
+  Add(equals_group, Wrap(nullptr), Wrap(nullptr));
+  Add(equals_group, Wrap(false), Wrap(false));
+  Add(equals_group, Wrap(true), Wrap(true));
+  Add(equals_group, Wrap(std::numeric_limits<double>::quiet_NaN()),
+      Wrap(ToDouble(kCanonicalNanBits)), Wrap(ToDouble(kNanBits)),
+      Wrap(std::nan("1")), Wrap(std::nan("2")));
+  // -0.0 and 0.0 compare the same but are not equal.
+  Add(equals_group, Wrap(-0.0));
+  Add(equals_group, Wrap(0.0));
+  Add(equals_group, Wrap(1), Wrap(1LL));
+  // Doubles and Longs aren't equal (even though they compare same).
+  Add(equals_group, Wrap(1.0), Wrap(1.0));
+  Add(equals_group, Wrap(1.1), Wrap(1.1));
+  Add(equals_group, Wrap(BlobValue(0, 1, 1)));
+  Add(equals_group, Wrap(BlobValue(0, 1)));
+  Add(equals_group, Wrap("string"), Wrap("string"));
+  Add(equals_group, Wrap("strin"));
+  // latin small letter e + combining acute accent
+  Add(equals_group, Wrap("e\u0301b"));
+  // latin small letter e with acute accent
+  Add(equals_group, Wrap("\u00e9a"));
```

Comment:
Other possible case: string with '\0' in the middle of it.

Reply:
There should already be done in the line below (added in that review cycle). I added additional ones for comparison + canoncial ID now.